src/optimizer/query_to_operator_transformer.cpp

```diff
@@ -31,113 +32,100 @@ using std::shared_ptr;
 namespace peloton {
 namespace optimizer {
 QueryToOperatorTransformer::QueryToOperatorTransformer(
-  concurrency::TransactionContext *txn)
-  : txn_(txn) {}
-
-
+    concurrency::TransactionContext *txn)
+    : txn_(txn),
+      get_id(0),
+      enable_predicate_push_down_(settings::SettingsManager::GetBool(
+          settings::SettingId::predicate_push_down)) {}
 std::shared_ptr<OperatorExpression>
 QueryToOperatorTransformer::ConvertToOpExpression(parser::SQLStatement *op) {
-  output_expr = nullptr;
+  output_expr_ = nullptr;
   op->Accept(this);
-  return output_expr;
+  return output_expr_;
 }
 
 void QueryToOperatorTransformer::Visit(parser::SelectStatement *op) {
-  auto upper_expr = output_expr;
-
-  if (op->where_clause != nullptr) {
-    SingleTablePredicates where_predicates;
-    util::ExtractPredicates(op->where_clause.get(), where_predicates, join_predicates_);
-
-    // Remove join predicates in the where clause
-    if (!join_predicates_.empty()) {
-      // Discard the const qualifier to update the where clause.
-      // The select statement can only be updated in this way. This is a hack.
-      ((parser::SelectStatement *)op)
-          ->UpdateWhereClause(util::CombinePredicates(where_predicates));
-    }
-    else {
-      for (auto expr : where_predicates)
-        delete expr;
-    }
-  }
+  auto pre_predicates = std::move(predicates_);
+  auto pre_depth = depth_;
+  depth_ = op->depth;
 
   if (op->from_table != nullptr) {
     // SELECT with FROM
     op->from_table->Accept(this);
-    if (op->group_by != nullptr) {
-      // Make copies of groupby columns
-      vector<shared_ptr<expression::AbstractExpression>> group_by_cols;
-      for (auto& col : op->group_by->columns)
-        group_by_cols.emplace_back(col->Copy());
-      auto group_by = std::make_shared<OperatorExpression>(
-          LogicalGroupBy::make(move(group_by_cols), op->group_by->having.get()));
-      group_by->PushChild(output_expr);
-      output_expr = group_by;
+  } else {
+    // SELECT without FROM
+    output_expr_ = std::make_shared<OperatorExpression>(LogicalGet::make());
+  }
+
+  if (op->where_clause != nullptr) {
+    CollectPredicates(op->where_clause.get());
+  }
+
+  if (!predicates_.empty()) {
+    auto filter_expr =
+        std::make_shared<OperatorExpression>(LogicalFilter::make(predicates_));
+    filter_expr->PushChild(output_expr_);
+    output_expr_ = filter_expr;
+  }
+
+  if (RequireAggregation(op)) {
+    // Plain aggregation
+    std::shared_ptr<OperatorExpression> agg_expr;
+    if (op->group_by == nullptr) {
+      agg_expr = std::make_shared<OperatorExpression>(LogicalGroupBy::make());
```

Comment:
I think this is for plain aggregation like "SELECT count(*) FROM table" where there's aggregation but no GROUP BY.

Reply:
But then why we are still construct a `LogicalGroupBy` here?

Reply:
Because we do not have `LogicalAggregate` operator. We transform `LogicalGroupBy` with empty group-by columns to `PhysicalAggregate`. Do you think we should add the logical operator?

Reply:
Eh... It's fine to have one logical operator, but then you should rename it to `LogicalAggregateAndGroupBy` and document this in the comment.